prow/plugins/owners/owners.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package owners
+
+import (
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+	"k8s.io/test-infra/prow/repoowners"
+)
+
+const (
+	pluginName     = "owners"
+	ownersFileName = "OWNERS"
+)
+
+var (
+	invalidOwnersLabel           = "do-not-merge/invalid-OWNERS"
+)
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	return &pluginhelp.PluginHelp{
+			Description: "The owners plugin validates formatting and 'labels' sections of OWNERS files modified in a PR. On failure it automatically adds a 'do-not-merge/invalid-OWNERS' label to the PR.",
+		},
+		nil
+}
+
+type ownersClient interface {
+	LoadRepoOwners(org, repo, base string) (repoowners.RepoOwnerInterface, error)
+}
+
+type githubClient interface {
+	AddLabel(owner, repo string, number int, label string) error
+	CreateComment(owner, repo string, number int, comment string) error
+	GetPullRequestChanges(org, repo string, number int) ([]github.PullRequestChange, error)
+}
+
+func handlePullRequest(pc plugins.PluginClient, pre github.PullRequestEvent) error {
+	if pre.Action != github.PullRequestActionOpened && pre.Action != github.PullRequestActionReopened && pre.Action != github.PullRequestActionSynchronize {
+		return nil
+	}
+	return handle(pc.GitHubClient, pc.OwnersClient, pc.Logger, &pre, pc.PluginConfig.Owners.LabelsBlackList)
+}
+
+func handle(ghc githubClient, oc ownersClient, log *logrus.Entry, pre *github.PullRequestEvent, labelsBlackList []string) error {
+	var errMsg error
+	changes, err := ghc.GetPullRequestChanges(pre.Repo.Owner.Login, pre.Repo.Name, pre.Number)
+	if err != nil {
+		return fmt.Errorf("error getting PR changes: %v", err)
+	}
+	for _, change := range changes {
+		if change.Filename == ownersFileName {
+			r, err := oc.LoadRepoOwners(pre.Repo.Owner.Login, pre.Repo.Name, pre.PullRequest.Base.Ref)
```

Comment:
This loads the OWNERS file data from a commit on the base branch (e.g. master), not the PR.
This plugin should checkout and merge the PR using Prow's [`git`](https://github.com/kubernetes/test-infra/tree/master/prow/git) package. Here is an example of that: https://github.com/kubernetes/test-infra/blob/3e8c8db547549298039a7ccd5069dd85b98d5446/prow/tide/tide.go#L801-L830

Then it can read and validate just the OWNERS files that the PR changes, there shouldn't be a need to load the `RepoOwners` struct because OWNERS files don't have interdependencies for us to validate.

Reply:
OWNERS files are already loaded into structs in the `repoowners` package. If you move that logic into a public function and also make the struct that represents individual OWNERS files public, the code can be shared with this plugin.

Reply:
thanks, I am correcting these points...